src/app/navigation/types.ts

```diff
@@ -1,14 +1,28 @@
 import { Location } from 'history';
+import { routes } from '../';
 
 export type State = Location;
 
-export interface Match<Params> {
-  route: Route<Params>;
-  params: Params;
+type RouteParams<R> = R extends Route<infer P> ? P : never;
+type UnionRouteMatches<R> = R extends AnyRoute ? Match<R> : never;
+
+interface MatchWithParams<R extends AnyRoute> {
```

Comment:
Ah nice! At first glance I was wondering why anyone would extend `AnyRoute`... but this way `route` and `params` have the *same* `AnyRoute`.